### PR TITLE
dirtools: handle empty symlink targets

### PR DESCRIPTION
**Pull request by sluongng**
Bazel can encode `target_path = "."` as a `SymlinkNode` whose target is the
proto3 default empty string. The remote execution proto describes symlink
target strings, but it does not define whether a local relative self-link
should be encoded as `"."` or as that empty normalized form.

In JSON, the input tree symlink looks like this because proto3 omits the
empty target:

```json
{
  "symlinks": [
    {
      "name": "PROJ"
    }
  ]
}
```

Our Executor previously passed `SymlinkNode.target` straight to `os.Symlink`
when materializing inputs. On Linux, `os.Symlink("", ".../PROJ")` fails
with `no such file or directory`, which surfaces as:

```text
download inputs: could not start tree fetcher
```

This PR first captures the failure, then fixes it by treating an empty remote
symlink target as `"."` while downloading inputs. The newest commit keeps the
same regression coverage but moves it to `server/cache/dirtools`: `DownloadTree`
now receives the malformed `SymlinkNode{Name: "PROJ"}` message directly and
verifies the materialized link points at `"."`.

Validated with:

```text
./buildfix.sh
bazel test --config=remote-minimal //server/cache/dirtools:dirtools_test
```

### CODE REVIEW

**Comment by sluongng** [comment]
reported by one of our prospects

**Comment by fmeum** [comment]
Even though it's a bit weird, Bazel's behavior makes sense as it ensures that there is a unique representation for each relative path. Which makes this a good fix and probably worth a remote apis spec update. We could bring this up in the next meeting.

File: server/cache/dirtools/dirtools.go [hunk 1]
```diff
@@ -1312,14 +1312,20 @@ type symlinkRequest struct {
 }
 
 func (slr *symlinkRequest) Do() error {
-	err := os.Symlink(slr.oldname, slr.newname)
+	oldname := slr.oldname
+	if oldname == "" {
+		// Bazel can encode target_path = "." as an empty REAPI SymlinkNode
```
Comment (fmeum) [inline server/cache/dirtools/dirtools.go:1317]: I think that it would currently always do this due to how `PathFragment` is implemented.

**Review by fmeum — APPROVED** [review]

File: enterprise/server/test/integration/remote_execution/bazel_rbe/bazel_rbe_test.go [hunk 2]
```diff
@@ -324,6 +324,67 @@ func TestTransientCacheNotFoundError_Retried(t *testing.T) {
 	assert.Equal(t, 2, GetNumExecutionsFlushedToOLAPDB(t, env))
 }
 
+func TestGeneratedSymlinkToCurrentDirectoryAsRemoteInput(t *testing.T) {
```
Comment (bduffany) [inline enterprise/server/test/integration/remote_execution/bazel_rbe/bazel_rbe_test.go]: These tests are pretty heavyweight - seems a tiny bit overkill to test this one edge case using a full integration test (vs a dirtools unit test)?

Comment (sluongng) [reply enterprise/server/test/integration/remote_execution/bazel_rbe/bazel_rbe_test.go]: Agree. The bot actually went with a unit test at first but i forced it to go with an integration test so i can verify how it works with Bazel.

can turn it back to a unit test now that we got that covered

File: server/cache/dirtools/dirtools.go (hunk 1 above)
Comment (sluongng) [reply server/cache/dirtools/dirtools.go:1317]:
```
  The path is:

  - ctx.actions.symlink(..., target_path = ".") stores ".", but UnresolvedSymlinkAction#getTargetPathFragment() calls
  PathFragment.create(target), with an existing TODO noting this normalizes symlink targets: /Users/sluongng/work/bazelbuild/
  bazel/src/main/java/com/google/devtools/build/lib/analysis/starlark/UnresolvedSymlinkAction.java:112
  - PathFragment.create(".") normalizes "." to the empty path: /Users/sluongng/work/bazelbuild/bazel/src/main/java/com/google/
  devtools/build/lib/vfs/PathFragment.java:123
  - Local symlink creation still works because JavaIoFileSystem#createSymbolicLink uses getSafePathString(), which maps empty
  back to ".": /Users/sluongng/work/bazelbuild/bazel/src/main/java/com/google/devtools/build/lib/vfs/JavaIoFileSystem.java:282
  - Later, when building the remote Merkle tree, Bazel reads the symlink with readSymbolicLink().getPathString(): /Users/
  sluongng/work/bazelbuild/bazel/src/main/java/com/google/devtools/build/lib/remote/merkletree/DirectoryTreeBuilder.java:190
  - That string is serialized directly as REAPI SymlinkNode.target: /Users/sluongng/work/bazelbuild/bazel/src/main/java/com/
  google/devtools/build/lib/remote/merkletree/MerkleTree.java:419
```

from the bot. Seems accurate, as we can reproduce it using the integration test earlier.
